### ISSUE TITLE
Web Cmdlet Socket Exhaustion Concern

### ISSUE THREAD

**Issue by Francisco-Gamino** [PowerShell/PowerShell#9977]
This issue was originally opened in the https://github.com/Azure/azure-functions-powershell-worker repo by @markekraus. 

One thing we do not handle well in the web cmdlets in PS Core is the singleton nature of HttpClient. HttpClient is intended to be instantiated once and then reused. In PS Core we instantiate a new HttpClient on every call to IWR/IRM. This is somewhat OK for a shell and scripting language, but could be problematic for Functions. A common pattern is to create a azure Function to combine data from multiple endpoints as a single data set. In a high volume environment, the repeated calls to web cmdlets could result in socket exhaustion. [More Info](https://docs.microsoft.com/en-us/aspnet/web-api/overview/advanced/calling-a-web-api-from-a-net-client#create-and-initialize-httpclient)

For conversation details please see https://github.com/Azure/azure-functions-powershell-worker/issues/174

**Comment by vexx32** [comment]
From what I've read on the subject, the fix could more or less be as simple as maintaining a static `HttpClient` instance that is reused for all web cmdlet calls to the API. If thread safety is a concern, locks or possibly [ThreadStatic] members may be needed as well.

Mark may have other ideas as well. 🙂

**Comment by markekraus** [comment]
Unfortunately, it's not that simple. lets say the following 2 commands run in different runspaces:

```powershell
Invoke-WebRequest -SkipCertificateCheck -Uri https://myinternaldevsite.com
Invoke-WebRequest -Uri https://sensitiveexternalsite.com
```

Because both of those make a modification to the `HttpClientHandler.ServerCertificateCustomValidationCallback` (which would need to be static between threads to reduce the risk of port exhaustion) it is possible that the `HttpClientHandler.DangerousAcceptAnyServerCertificateValidator` could be applied when the call to `https://sensitiveexternalsite.com` is made.

Basically, there are a few options. 1 is to create an `HttpClient` and `HttpClientHandler` pair per thread. Another is to implement our own `HttpMessageHandler` that includes the logic for handling our use cases (this is a pretty massive undertaking). Another is to create some kind of smart logic to create a new `HttpClient` only when needed (not as easy as it sounds).

**Comment by vexx32** [comment]
.NET Core uses an implementation of a StringBuilderCache that is similar to what it sounds like you're suggesting; we could possibly take some pointers from their implementation of that and go from there....

**Comment by JustinGrote** [comment]
Since there is a fairly limited amount of HttpClientHandler configurations, and it's unlikely a user is going to  instantiate multiple different clienthandler configurations per session (maybe 2-3 to accomodate different host requirements), as discussed on another thread it's probably best to instantiate one httpclient per httpclienthandler configuration, which will drastically reduce the number of httpclients instantiated and bring the socket count down, especially for multiple concurrent requests to the same host using the same httpclienthandler configuration.

**Comment by iSazonov** [comment]
I see that @daxian-dbw suggest per runspace caching https://github.com/Azure/azure-functions-powershell-worker/issues/174#issuecomment-476791645
It is easy to check that it improves performance without feature regression.

**Comment by JustinGrote** [comment]
The problem is that there still needs to allow multiple httpclienthandlers for different iwr parameter configurations, and it's a 1:1 relationship for httpclienthandler to httpclient.

Not sure what he means by the race conditions either, httpclient is thread safe

**Comment by iSazonov** [comment]
> for these reasons I think a process wide singleton per httpclienthandler would be sufficient.

So we need cache like config -> httpclienthandler or config -> httpclienthandler -> httpclient

**Comment by JustinGrote** [comment]
I'll defer to @markekraus on the particular implementation, I'm not a c# wizard but I'd envision taking a hash of the config and using it as the key in a key value cache, the value being the httpclient that was constructed with the particular httpclienthandler, and instantiate a new one.

The cache would dispose with the process, since there's a fairly limited group of configurations especially in a single use case so a memory leak isn't a major concern.

**Comment by iSazonov** [comment]
We could use a (global static) cache like Tokenizer.GetStringBuilder().

**Comment by vexx32** [comment]
Not a bad idea. It's probably OK to just have the one handler to start off, and if the user requests a different configuration, the method goes off to create the necessary handler, store it in the lookup, and then return it.

**Comment by iSazonov** [comment]
For reference https://github.com/dotnet/runtime/issues/44818 - what we could cache.

**Comment by CarloToso** [comment]
@iSazonov Duplicate of #13109 and #6156 related to #12764 might be fixed by #19249

**Comment by iSazonov** [comment]
The issue resolved if WebSession is used but not otherwise.
But I doubt that it makes sense to cache the session in the shadow always because requests can be to different servers and with different parameters, and it can be too expensive to create a cache for each target server.

**Comment by JustinGrote** [comment]
Don't all httpclients now just use socketshttphandler that coordinates for socket exhaustion now since .net core 2.1? If so then the default behavior shouldn't really be a worry anymore, unless there's special handling going on.

**Comment by iSazonov** [comment]
Yes. And now (7.4 Preview2) we explicitly dispose HttpClient and HttpClientHandler/SocketsHttpHandler.

**Comment by JustinGrote** [comment]
@Francisco-Gamino I would argue this can be closed as resolved then
